packages/liveblocks-react/src/comments/CommentsRoom.ts

```diff
@@ -114,732 +118,577 @@ export type ThreadsStateSuccess<TThreadMetadata extends BaseMetadata> = {
-    },
-  };
-}
-
-function handleCommentsApiError(err: CommentsApiError): Error {
-  const message = `Request failed with status ${err.status}: ${err.message}`;
-
-  // Log details about FORBIDDEN errors
-  if (err.details?.error === "FORBIDDEN") {
-    const detailedMessage = [message, err.details.suggestion, err.details.docs]
-      .filter(Boolean)
-      .join("\n");
-
-    console.error(detailedMessage);
-  }
-
-  return new Error(message);
-}
-
-/**
- * This implementation is inspired by the `swr` library.
- * Additional modifications were made to adapt it to our specific needs.
- *
- * Original `swr` library can be found at [SWR GitHub repository](https://github.com/vercel/swr)
- */
 export function createCommentsRoom<TThreadMetadata extends BaseMetadata>(
   room: Room<JsonObject, LsonObject, BaseUserMeta, Json>,
   errorEventSource: EventSource<CommentsError<TThreadMetadata>>
 ): CommentsRoom<TThreadMetadata> {
-  const manager = createThreadsManager<TThreadMetadata>();
+  const manager = createCacheManager<ThreadData<TThreadMetadata>[]>();
 
-  let timestamp = 0;
+  function _useThreads(
+    revalidateCache: (shouldDedupe: boolean) => Promise<void>
+  ): ThreadsState<TThreadMetadata> {
+    const status = useSyncExternalStore(
+      room.events.status.subscribe,
+      room.getStatus,
+      room.getStatus
+    );
 
-  let commentsEventRefCount = 0; // Reference count for the number of components with a subscription (via the `subscribe` function)
-  let commentsEventDisposer: (() => void) | undefined; // Disposer function for the `comments` event listener
+    const interval =
+      status === "connected" ? POLLING_INTERVAL_REALTIME : POLLING_INTERVAL;
 
-  async function mutate(
-    data: Promise<any>,
-    options: {
-      optimisticData: ThreadData<TThreadMetadata>[];
-    }
-  ) {
-    const beforeMutationTimestamp = ++timestamp;
-    manager.mutation = {
-      startTime: beforeMutationTimestamp,
-      endTime: 0,
-    };
+    // Automatically revalidate the cache when the window gains focus or when the connection is restored. Also poll the server based on the connection status.
+    useAutomaticRevalidation(manager, revalidateCache, {
+      revalidateOnFocus: true,
+      revalidateOnReconnect: true,
+      refreshInterval: interval,
+    });
 
-    const currentCache = manager.cache;
+    /**
+     * Subscribe to comment events in the room to trigger a revalidation when a comment is added, edited or deleted.
+     */
+    useEffect(() => {
+      const unsubscribe = room.events.comments.subscribe(() => {
+        void revalidateCache(true);
+      });
+      return () => {
+        unsubscribe();
+      };
+    }, [revalidateCache]);
 
-    // Update the cache with the optimistic data
-    manager.cache = {
-      isLoading: false,
-      threads: options.optimisticData,
-    };
+    const cache = useSyncExternalStore(
+      manager.subscribe,
+      () => manager.cache,
+      () => manager.cache
```

Comment:
Should we re-use those and more importantly `useCallback` it?

Reply:
1️⃣ I think it might be okay for now to keep it inline for better readability despite the repetition. Also, the repeated code snippet is quite small, but might be worth keeping an eye on in future. Do you have preference for one over the other? 

2️⃣ I believe it is fine without memoizing. From what I understand, the value returned by selector function in `useSyncExternalStore` is what is required to be memoized or cached, not necessarily the function itself.

https://react.dev/reference/react/useSyncExternalStore#im-getting-an-error-the-result-of-getsnapshot-should-be-cached
> React will re-render the component if getSnapshot return value is different from the last time. This is why, if you always return a different value, you will enter an infinite loop and get this error.


https://github.com/reactwg/react-18/discussions/86
> Selectors no longer need to be memoized
>
> Inline selectors are a common feature of state libraries, especially those that provide a hook-based API, such as Redux’s useSelector.

Reply:
I tend to be overzealous with these things (cc: @nvie), you're right it's fine as is 👍